bson/vector.go

```diff
@@ -0,0 +1,271 @@
+	if !ok {
+		panic(vectorTypeError{"bson.Vector.Int8", v.dType})
+	}
+	return d
+}
+
+// Int8OK is the same as Int8, but returns a boolean instead of panicking.
+func (v Vector) Int8OK() ([]int8, bool) {
+	if v.dType != Int8Vector {
+		return nil, false
+	}
+	return v.int8Data, true
+}
+
+// Float32 returns the float32 slice hold by the vector.
+// It panics if v is not a float32 vector.
+func (v Vector) Float32() []float32 {
+	d, ok := v.Float32OK()
+	if !ok {
+		panic(vectorTypeError{"bson.Vector.Float32", v.dType})
+	}
+	return d
+}
+
+// Float32OK is the same as Float32, but returns a boolean instead of panicking.
+func (v Vector) Float32OK() ([]float32, bool) {
+	if v.dType != Float32Vector {
+		return nil, false
+	}
+	return v.float32Data, true
+}
+
+// PackedBit returns the byte slice representing the binary quantized (packed bit) vector and the byte padding, which
+// is the number of bits in the final byte that are to be ignored.
+// It panics if v is not a packed bit vector.
+func (v Vector) PackedBit() ([]byte, uint8) {
+	d, p, ok := v.PackedBitOK()
+	if !ok {
+		panic(vectorTypeError{"bson.Vector.PackedBit", v.dType})
+	}
+	return d, p
+}
+
+// PackedBitOK is the same as PackedBit, but returns a boolean instead of panicking.
+func (v Vector) PackedBitOK() ([]byte, uint8, bool) {
+	if v.dType != PackedBitVector {
+		return nil, 0, false
+	}
+	return v.bitData, v.bitPadding, true
+}
+
+// Binary returns the BSON Binary of the Vector.
+func (v Vector) Binary() Binary {
+	switch v.Type() {
+	case Int8Vector:
+		return binaryFromInt8Vector(v.Int8())
+	case Float32Vector:
+		return binaryFromFloat32Vector(v.Float32())
+	case PackedBitVector:
+		return binaryFromBitVector(v.PackedBit())
+	default:
+		panic("invalid Vector type")
+	}
+}
+
+func binaryFromInt8Vector(v []int8) Binary {
+	data := make([]byte, 2, len(v)+2)
+	copy(data, []byte{byte(Int8Vector), 0})
+	for _, e := range v {
+		data = append(data, byte(e))
+	}
+
+	return Binary{
+		Subtype: TypeBinaryVector,
+		Data:    data,
+	}
+}
+
+func binaryFromFloat32Vector(v []float32) Binary {
+	data := make([]byte, 2, len(v)*4+2)
+	copy(data, []byte{byte(Float32Vector), 0})
+	var a [4]byte
+	for _, e := range v {
+		binary.LittleEndian.PutUint32(a[:], math.Float32bits(e))
+		data = append(data, a[:]...)
+	}
+
+	return Binary{
+		Subtype: TypeBinaryVector,
+		Data:    data,
+	}
+}
+
+func binaryFromBitVector(bits []byte, padding uint8) Binary {
+	data := []byte{byte(PackedBitVector), padding}
+	data = append(data, bits...)
+	return Binary{
+		Subtype: TypeBinaryVector,
+		Data:    data,
+	}
+}
+
+// NewVector constructs a Vector from a slice of int8 or float32.
+func NewVector[T int8 | float32](data []T) Vector {
+	var v Vector
+	switch a := any(data).(type) {
+	case []int8:
+		v.dType = Int8Vector
+		v.int8Data = []int8{}
+		v.int8Data = append(v.int8Data, a...)
+	case []float32:
+		v.dType = Float32Vector
+		v.float32Data = []float32{}
+		v.float32Data = append(v.float32Data, a...)
```

Comment:
Allocate the exact data size for the copy.

```suggestion
		v.float32Data = make([]float32, len(data))
		copy(v.float32Data, a)
```